### PR TITLE
Updating Oauth application owner

### PR DESCRIPTION
**Pull request by Ambujaa10**

### CODE REVIEW

**Comment by menakaj** [comment]
Merge the last 2 commits (Refactoring) into one commit and add a meaningful commit message.

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIAdmin.java [hunk 1]
```diff
@@ -60,4 +60,32 @@
...
+     * @param start                content to start
+     * @param offset               content to limit number of pages
+     * @param searchOwner          content to search applications based on owners
+     * @param searchApplication    content to search applications based on application
+     * @param sortColumn           content to sort column
+     * @param sortOrder            content to sort in a order
+     * @throws APIManagementException if failed to get applications
+     */
+    List<Application> getApplicationsByTenantIdWithPagination(int tenantId, int start , int offset
+            , String searchOwner, String searchApplication, String sortColumn, String sortOrder)
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIAdmin.java]: Fix the formatting issue. ','

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java [hunk 2]
```diff
@@ -208,6 +208,34 @@ public OAuthApplicationInfo updateApplication(OAuthAppRequest appInfoDTO) throws
...
+            org.wso2.carbon.apimgt.api.model.xsd.OAuthApplicationInfo applicationInfo = updateOAuthApplicationOwner(
+                    oAuthApplicationInfo.getAppOwner(), owner, applicationName,
+                    oAuthApplicationInfo.getCallBackURL(),oAuthApplicationInfo.getClientId(), grantTypes);
+            OAuthApplicationInfo newAppInfo = new OAuthApplicationInfo();
+            newAppInfo.setAppOwner(applicationInfo.getAppOwner());
+            newAppInfo.setClientId(applicationInfo.getClientId());
+            newAppInfo.setCallBackURL(applicationInfo.getCallBackURL());
+            newAppInfo.setClientSecret(applicationInfo.getClientSecret());
+            newAppInfo.setJsonString(applicationInfo.getJsonString());
+
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java]: This blank line is un-necessary

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java [hunk 3]
```diff
@@ -709,6 +737,20 @@ public AccessTokenInfo getAccessTokenByConsumerKey(String consumerKey) throws AP
...
+            String[] grantTypes) throws Exception {
+        SubscriberKeyMgtClient keyMgtClient = null;
+        try {
+            keyMgtClient = SubscriberKeyMgtClientPool.getInstance().get();
+            return keyMgtClient
+                    .updateOAuthApplicationOwner(userId, owner, applicationName, callBackURL, clientId, grantTypes);
+        } finally {
+            SubscriberKeyMgtClientPool.getInstance().release(keyMgtClient);
+        }
+
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java]: Remove the blank line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 4]
```diff
@@ -4408,6 +4409,94 @@ public int compare(Application o1, Application o2) {
         return applications;
     }
 
+    /**
+     *
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Add a proper Javadoc comment explaining the functionality.

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyMgtSubscriberService.java [hunk 5]
```diff
@@ -490,6 +486,143 @@ public OAuthApplicationInfo updateOAuthApplication(String userId, String applica
...
+                            if ("authorization_code".equals(grantType) || "implicit".equals(grantType)) {
+                                continue;
+                            }
+                        }
+                        grantTypeString.append(grantType).append(" ");
+                    }
+                    oAuthConsumerAppDTO.setGrantTypes(grantTypeString.toString().trim());
+                }
+
+
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyMgtSubscriberService.java]: Remove extra lines.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/modules/application-owner/change-owner.jag [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    } catch (ex) {
+        var errorMessage = ex.message.split(":")[1];
+        log.error(errorMessage);
+        return {
+            error: true,
+            message: errorMessage
+        };
+    }        
+};
+%>
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/modules/application-owner/change-owner.jag]: Add a new line at the end of the file

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+         element.css("border", "1px solid red");
+         $('#label'+elementId).remove();
+         element.parent().append('<label class="error" id="label'+elementId+'" >' + errorMsg + '</label>');
+         return false;
+     }else{
+         $('#label'+elementId).remove();
+         element.css("border", "1px solid #cccccc");
+         return true;
+     }
+ }
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js]: Add new line

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/get-applications/js/get-applications.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+                {"data": "name"},
+                {"data": "owner"},
+                {"data": "name", "render": function ( data, type, rec, meta ) {
+                  return application_actions(rec);
+              }
+            }
+            ]
+        });
+    });
+});
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/get-applications/js/get-applications.js]: Add a new line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 9]
```diff
@@ -4057,56 +4059,60 @@ public void deleteOAuthApplication(String consumerKey) throws APIManagementExcep
...
-
-        } catch (org.wso2.carbon.user.api.UserStoreException e) {
-            handleException("Error when getting the tenant's UserStoreManager or when getting roles of user ", e);
-        } catch (RemoteException e) {
-            handleException("Server couldn't establish connection with auth manager ", e);
-        } catch (UserAdminUserAdminException e) {
-            handleException("Error when getting the tenant's UserStoreManager or when getting roles of user ", e);
+        } else {
+            throw new APIManagementException("Unable to update application owner to " +
+                    userId + " as this user does not belong to this tenant domain");
```
Comment (sameeragunarathne) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: shouldn't we add the tenant domain information to the log message?

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyMgtSubscriberService.java [hunk 10]
```diff
@@ -490,6 +486,143 @@ public OAuthApplicationInfo updateOAuthApplication(String userId, String applica
...
+                    builder.deleteCharAt(builder.length() - 1);
+                    oAuthConsumerAppDTO.setGrantTypes(builder.toString());
+                } else {
+                    //update the grant type with respect to callback url
+                    String[] allowedGrantTypes = oAuthAdminService.getAllowedGrantTypes();
+                    StringBuilder grantTypeString = new StringBuilder();
+
+                    for (String grantType : allowedGrantTypes) {
+                        if (callbackUrl == null || callbackUrl.isEmpty()) {
+                            if ("authorization_code".equals(grantType) || "implicit".equals(grantType)) {
```
Comment (sameeragunarathne) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyMgtSubscriberService.java]: can we add constants for these strings?

Comment (Ambujaa10) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyMgtSubscriberService.java]: Constants are added for the strings

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 9 above)
Comment (Ambujaa10) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: Tenant domain information is added

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/get-applications/js/get-applications.js (hunk 8 above)
Comment (Ambujaa10) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/get-applications/js/get-applications.js]: Done

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js (hunk 7 above)
Comment (Ambujaa10) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js]: Done

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/modules/application-owner/change-owner.jag (hunk 6 above)
Comment (Ambujaa10) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/modules/application-owner/change-owner.jag]: Done

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyMgtSubscriberService.java (hunk 5 above)
Comment (Ambujaa10) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyMgtSubscriberService.java]: Extra lines are removed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 4 above)
Comment (Ambujaa10) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Javadoc comment is added

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java (hunk 3 above)
Comment (Ambujaa10) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java]: Done

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java (hunk 2 above)
Comment (Ambujaa10) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java]: Done

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIAdmin.java (hunk 1 above)
Comment (Ambujaa10) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIAdmin.java]: Code formatted

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/application-owner/get-applications/ajax/get-applications.jag [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        var apiAdmin = new APIAdminImpl();
+        if (sortOrder.toLowerCase() != "asc" && sortOrder.toLowerCase() != "desc" ) {
+        print({
+            error:true,
+            message:msg.error.forbidAction(action)
+            });
+            return;
+        }
+        var sortColumn = null;
+        if(columnId == 0){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/application-owner/get-applications/ajax/get-applications.jag]: Please fix the formatting issues.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        element.css("border", "1px solid #cccccc");
+        return true;
+    }
+}
+
+function validateInputCharactors(text, element, errorMsg){
+    var elementId = element.attr('id');
+    var illegalChars = /([~!&@#;%^*+={}\|\\<>\"\',])/;     
+    text = text.trim();
+    if(illegalChars.test(text)){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js]: Formatting issue.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+                 var attributeName = attributeElement.val();
+                 var attributeValue = attributeValueElement.val();
+ 
+                 if(!validateAttributesInput(attributeName, attributeElement, requiredMsg, invalidErrorMsg)){
+                     isInvalidAttribute = true;
+                     return false;
+                 }
+                 // We do not validate the attribute value input as it can be empty.
+             });
+             if(isInvalidAttribute){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js]: formatting issue.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/template.jag [hunk 14]
```diff
@@ -0,0 +1,71 @@
+<% jagg.template("application-owner/change-owner", function(inputs, outputs, jagg) { %>
+
+<%
+    var applicationUuid = request.getParameter("uuid");
+    var applicationAction = request.getParameter("action");
+    var applicationName = request.getParameter("name");
+    var applicationOwner = request.getParameter("owner");
+    var nameFieldDisableStatus = "";
+
+    if(applicationUuid){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/template.jag]: Formatting issue.

Comment (Ambujaa10) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/template.jag]: Done

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js (hunk 13 above)
Comment (Ambujaa10) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js]: Done

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js (hunk 12 above)
Comment (Ambujaa10) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/application-owner/change-owner/js/change-owner.js]: Done

**Pull request merged**